gcloud-java-dns/src/main/java/com/google/gcloud/dns/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1266 @@
+
+    ZoneContainer(ManagedZone zone) {
+      this.zone = zone;
+      this.dnsRecords.set(ImmutableSortedMap.<String, ResourceRecordSet>of());
+    }
+
+    ManagedZone zone() {
+      return zone;
+    }
+
+    AtomicReference<ImmutableSortedMap<String, ResourceRecordSet>> dnsRecords() {
+      return dnsRecords;
+    }
+
+    ConcurrentLinkedQueue<Change> changes() {
+      return changes;
+    }
+
+    Change findChange(String changeId) {
+      for (Change current : changes) {
+        if (changeId.equals(current.getId())) {
+          return current;
+        }
+      }
+      return null;
+    }
+  }
+
+  static class Response {
+    private final int code;
+    private final String body;
+
+    Response(int code, String body) {
+      this.code = code;
+      this.body = body;
+    }
+
+    int code() {
+      return code;
+    }
+
+    String body() {
+      return body;
+    }
+  }
+
+  private enum Error {
+    REQUIRED(400, "global", "required", "REQUIRED"),
+    INTERNAL_ERROR(500, "global", "internalError", "INTERNAL_ERROR"),
+    BAD_REQUEST(400, "global", "badRequest", "BAD_REQUEST"),
+    INVALID(400, "global", "invalid", "INVALID"),
+    CONTAINER_NOT_EMPTY(400, "global", "containerNotEmpty", "CONTAINER_NOT_EMPTY"),
+    NOT_AVAILABLE(400, "global", "managedZoneDnsNameNotAvailable", "NOT_AVAILABLE"),
+    NOT_FOUND(404, "global", "notFound", "NOT_FOUND"),
+    ALREADY_EXISTS(409, "global", "alreadyExists", "ALREADY_EXISTS"),
+    CONDITION_NOT_MET(412, "global", "conditionNotMet", "CONDITION_NOT_MET"),
+    INVALID_ZONE_APEX(400, "global", "invalidZoneApex", "INVALID_ZONE_APEX");
+
+    private final int code;
+    private final String domain;
+    private final String reason;
+    private final String status;
+
+    Error(int code, String domain, String reason, String status) {
+      this.code = code;
+      this.domain = domain;
+      this.reason = reason;
+      this.status = status;
+    }
+
+    Response response(String message) {
+      try {
+        return new Response(code, toJson(message));
+      } catch (IOException e) {
+        return Error.INTERNAL_ERROR.response("Error when generating JSON error response.");
+      }
+    }
+
+    private String toJson(String message) throws IOException {
+      Map<String, Object> errors = new HashMap<>();
+      errors.put("domain", domain);
+      errors.put("message", message);
+      errors.put("reason", reason);
+      Map<String, Object> args = new HashMap<>();
+      args.put("errors", ImmutableList.of(errors));
+      args.put("code", code);
+      args.put("message", message);
+      args.put("status", status);
+      return jsonFactory.toString(ImmutableMap.of("error", args));
+    }
+  }
+
+  private class RequestHandler implements HttpHandler {
+
+    private Response pickHandler(HttpExchange exchange, CallRegex regex) {
+      String path = BASE_CONTEXT.relativize(exchange.getRequestURI()).getPath();
```

Comment:
You call `BASE_CONTEXT.relativize(exchange.getRequestURI())` twice.  You can call once and save the result before calling `getPath()` and `getQuery()`